Introduce modularity

gberaudo opened the pull request:
This PR modularizes the ngeo search directive.
See https://github.com/camptocamp/ngeo/issues/2004 for the original issue and proposition.
The second commit is an improvement to the proposition.

The directive is made available to the library itself and to client applications through an Angular module.
This is on a par with the Angular guide available at https://docs.angularjs.org/guide/module.

This is a first step, the adaptation work will be incremental, depending on the needs and contributions of client applications.

Client applications interested by this approach includes Suissealpine and Geoadmin and should bring the following contributions:
- WMTS/WMS/GPX/... import module from Geoadmin;
- Swisstopo name component/directive for Suissealpine (ported from Schweizmobil);
- Swisstopo height directive for Suissealpine.

The long term goal is having the ngeo part of the library modularized.

sbrunner commented on src/modules/search/search.js:
```diff
@@ -1,6 +1,13 @@
+/**
+ * @module ngeo search namespace
+ */
+goog.provide('ngeo.modules.search');
```
Why not ngeo.search?

gberaudo replied:
In the prospect of migrating one day to ES6 modules I kept the naming of the filesystem.

sbrunner commented on src/modules/search/search.js:
```diff
@@ -24,7 +31,7 @@ goog.require('ngeo');
  * @ngname ngeoSearch
  */
-ngeo.searchDirective = function() {
+ngeo.modules.search.Directive = function() {
```
ngeo.search.Directive ?

sbrunner commented on src/ngeo.js:
```diff
@@ -20,9 +20,13 @@ goog.require('ol.layer.VectorTile');
 /** @type {!angular.Module} */
-ngeo.module = angular.module('ngeo', ['gettext', 'ui.date', 'floatThead']);
+ngeo.module = angular.module('ngeo', [
+    ngeo.modules.search.module.name,
```
I don't know about that ...

gberaudo replied:
This simply get the `ngeoSearch` string from the module.

sbrunner commented on examples/search.js:
```diff
@@ -17,7 +16,7 @@ goog.require('ol.source.Vector');
 
 /** @type {!angular.Module} **/
-app.module = angular.module('app', ['ngeo']);
+app.module = angular.module('app', [ngeo.module.name]);
```
Shouldn't depends on ngeo.serch module ?

gberaudo replied:
The example depended on the main ngeo module. The search module is already a dependency of the ngeo module so there is no need to add it here.

gberaudo commented:
@sbrunner I moved the creategeojsonbloodhound file.

gberaudo commented:
@sbrunner, as discussed I renamed the files, but without `ngeo` prefixes; it was making the names too ugly.
I added a README.md explaining the guidelines for this directory. Does it look good to you?

sbrunner commented on src/modules/README.md:
```diff
@@ -0,0 +1,19 @@
+
+
+# Google library
+The code do not use the Google library with exceptions for `goog.provide`, `goog.require` and `goog.asserts.*`.
```
:-)

gberaudo commented:
I relaxed the filename rules to make it more pleasant to use. As a consequence, it is not yet completly ES6 module friendly but I think it is better to go slowly than rush.

@sbrunner, @fredj could you please have a last look?

fredj commented on src/modules/search/creategeojsonbloodhound.js:
```diff
@@ -1,10 +1,10 @@
+/**
+ * @module ngeo search namespace
+ */
+goog.provide('ngeo.search.creategeojsonbloodhound');
```
name should be ngeo.search.createGeoJSONBloodhound

fredj commented on src/modules/search/creategeojsonbloodhound.js:
```diff
@@ -41,13 +41,13 @@ goog.require('ol.format.GeoJSON');
 
 /**
- * @param {string} url an URL to a search service.
+ * @param {string} url an URL to a search.creategeojsonbloodhound.
```
bad replace

gberaudo commented:
Thanks for the review, I handled all comments. Now waiting for the Travis green light.
